src/d.ts

```diff
@@ -36,6 +36,28 @@ export function isHNode(child: DNode): child is HNode {
 	return Boolean(child && (typeof child !== 'string') && child.type === HNODE);
 }
 
+/**
+ * Generic decorate function for DNodes, the modified in place based on the provided predicate
+ * and modifier functions.
+ *
+ * If no predicate is supplied then the modifier will be executed on all nodes.
+ */
+export function decorate(dNodes: DNode | DNode[], modifier: (dNode: DNode) => void, predicate?: (dNode: DNode) => void): DNode | DNode[] {
+	let nodes = Array.isArray(dNodes) ? [ ...dNodes ] : [ dNodes ];
+	while (nodes.length) {
+		const node = nodes.pop();
+		if (node) {
+			if (!predicate || predicate(node)) {
```

Comment:
The general concept is that the decorator accepts a single or array of `DNode`s, a `modifier` function and optionally a `predicate`.

It loops through each of the DNodes, checking the `predicate` by passing the current node and evaluating the return value (`true`/`false`) and if `true`, runs the node through the `modifier`. If there is no `predicate` provided then the `modifier` is run for all nodes.

As the decorator iterates through the nodes it spreads the children of each onto the node array in order to process them through the decorator, essentially flatten the tree structure of DNodes for decoration.

The original DNode argument (single or array) is returned (having had the modifications applied).